torch/optim/swa_utils.py

```diff
@@ -0,0 +1,207 @@
+        super(AveragedModel, self).__init__()
+        self.module = deepcopy(model)
+        if device is not None:
+            self.module = self.module.to(device)
+        self.register_buffer('n_averaged', torch.tensor(0, dtype=torch.long))
+        if avg_fun is None:
+            def avg_fun(p_avg, p, n_avg): 
+                return p_avg + (p - p_avg) / (n_avg + 1)
+        self.avg_fun = avg_fun
+
+    def forward(self, *args, **kwargs):
+        return self.module(*args, **kwargs)
+
+    def update_parameters(self, model):
+        for p_swa, p_model in zip(self.parameters(), model.parameters()):
+            device = p_swa.device
+            p_model_ = p_model.data.to(device)
+            if self.n_averaged == 0:
+                p_swa.data.copy_(p_model_)
+            else:
+                p_swa.data.copy_(self.avg_fun(p_swa.data, p_model_,
+                                              self.n_averaged))
+        self.n_averaged += 1
+
+
+def update_bn(loader, model, device=None):
+    r"""Updates BatchNorm running_mean, running_var buffers in the model.
+
+    It performs one pass over data in `loader` to estimate the activation
+    statistics for BatchNorm layers in the model.
+    Arguments:
+        loader (torch.utils.data.DataLoader): dataset loader to compute the
+            activation statistics on. Each data batch should be either a
+            tensor, or a list/tuple whose first element is a tensor
+            containing data.
+        model (torch.nn.Module): model for which we seek to update BatchNorm
+            statistics.
+        device (torch.device, optional): If set, data will be trasferred to
+            :attr:`device` before being passed into :attr:`model`.
+
+    Example:
+        >>> loader, model = ...
+        >>> torch.optim.swa_utils.update_bn(loader, model) 
+
+    .. note::
+        The `update_bn` utility assumes that each data batch in :attr:`loader`
+        is either a tensor or a list or tuple of tensors; in the latter case it 
+        is assumed that :meth:`model.forward()` should be called on the first 
+        element of the list or tuple corresponding to the data batch.
+    """
+    momenta = {}
+    has_bn = False
+    for module in model.modules():
+        if isinstance(module, torch.nn.modules.batchnorm._BatchNorm):
+            module.running_mean = torch.zeros_like(module.running_mean)
+            module.running_var = torch.ones_like(module.running_var)
+            momenta[module] = module.momentum
+            has_bn = True
+
+    if not has_bn:
+        return
+
+    was_training = model.training
+    model.train()
+    for module in momenta.keys():
+        module.momentum = None
+        module.num_batches_tracked *= 0
+
+    for input in loader:
+        if isinstance(input, (list, tuple)):
+            input = input[0]
+        if device is not None:
+            input = input.to(device)
+
+        model(input)
+
+    for module in model.modules():
```

Comment:
This loop can be simpler if done as above:
```
for bn in momenta.keys():
  bn.momentum = momenta[bn]
```

Reply:
thanks, fixed!